File: src/main/java/org/lanternpowered/server/entity/LanternHuman.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.entity;

import org.lanternpowered.server.data.LocalKeyRegistry;
import org.lanternpowered.server.data.key.LanternKeys;
import org.lanternpowered.server.inventory.vanilla.VanillaInventoryArchetypes;
import org.lanternpowered.server.network.entity.EntityProtocolTypes;
import org.lanternpowered.server.profile.LanternProfileProperty;
import org.spongepowered.api.data.Keys;
import org.spongepowered.api.entity.living.Human;
import org.spongepowered.api.item.inventory.Carrier;
import org.spongepowered.api.item.inventory.equipment.EquipmentInventory;
import org.spongepowered.api.item.inventory.type.CarriedInventory;
import org.spongepowered.api.text.Text;

import java.util.HashSet;
import java.util.UUID;

public class LanternHuman extends LanternAgent implements Human, AbstractArmorEquipable {

    private final EquipmentInventory equipmentInventory;

    public LanternHuman(UUID uniqueId) {
        super(uniqueId);
        setEntityProtocolType(EntityProtocolTypes.HUMAN);
        this.equipmentInventory = VanillaInventoryArchetypes.ENTITY_EQUIPMENT.build();
    }

    @Override
    public void registerKeys() {
        super.registerKeys();

        final LocalKeyRegistry<?> c = getKeyRegistry();
        c.register(LanternKeys.DISPLAYED_SKIN_PARTS, new HashSet<>());
        c.register(LanternKeys.POSE, Pose.STANDING);
        c.register(Keys.SKIN, LanternProfileProperty.EMPTY_TEXTURES);
    }

    @Override
    public String getName() {
        return get(Keys.DISPLAY_NAME).map(Text::toPlain).orElse("Unknown");
    }

    @Override
    public CarriedInventory<? extends Carrier> getInventory() {
        return this.equipmentInventory;
    }
}

File: src/main/kotlin/org/lanternpowered/server/entity/LanternHuman.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.entity

import org.lanternpowered.api.text.Text
import org.lanternpowered.api.text.toPlain
import org.lanternpowered.server.data.key.LanternKeys
import org.lanternpowered.server.inventory.equipment.AbstractArmorEquipable
import org.lanternpowered.server.inventory.vanilla.VanillaInventoryArchetypes
import org.lanternpowered.server.network.entity.EntityProtocolTypes
import org.lanternpowered.api.data.Keys
import org.lanternpowered.api.item.inventory.equipment.EquipmentInventory
import org.spongepowered.api.entity.living.Human
import java.util.UUID

class LanternHuman(creationData: EntityCreationData) : LanternAgent(creationData), Human, AbstractArmorEquipable {

    private val equipmentInventory: EquipmentInventory

    init {
        this.protocolType = EntityProtocolTypes.HUMAN
        this.equipmentInventory = VanillaInventoryArchetypes.ENTITY_EQUIPMENT.build()

        keyRegistry {
            register(LanternKeys.DISPLAYED_SKIN_PARTS, emptySet())
            register(LanternKeys.POSE, Pose.STANDING)
            register(Keys.SKIN_PROFILE_PROPERTY)
        }
    }

    override fun getName(): String = this.get(Keys.DISPLAY_NAME).map(Text::toPlain).orElse("Unknown")
    override fun getEquipment(): EquipmentInventory = this.equipmentInventory

    override fun useSkinFor(minecraftAccount: UUID): Boolean {
        TODO("Not yet implemented")
    }

    override fun useSkinFor(minecraftUsername: String?): Boolean {
        TODO("Not yet implemented")
    }
}
